# <font color=red>Solutions for</font> Exercises for Lab 03 <font color=blue>Version 1.1 </font>

## Functions used for Tables:
Just like in Lab 02 Exercises, the bulk of this first part is to provide a reference for the functions you'll use. The tables here are taken from the datascience module, so all these table functions can only work if you import datascience (see the first line of code after this).

Remember that in a given table, lists make up the rows and arrays make up columns. Additionally, tables have separate indexes for rows and columns. An index of 0 could refer to the first column or first row (the header row is not given an index) depending on the function. However, if you're typing `existingTable[index]`, it will go for columns, not rows.

I'm using the same conventions as I did in Exercises 02, namely that all parts that you should change as befit the situation you are in are given capital letters, and those that are to stay the same are all-lowercase. However, the Table datatype naturally has a capital T, and as such, whenever you see a `Table` *by itself*, you shouldn't change that.

I'm also including the `[etc,etc]` feature, that states that the inputs for these methods can stretch on and on to infinity (assuming that the relevant pattern is still fulfilled). For example, the .with_columns method can add on infinitely many columns so long as the pattern of Header then Array is fulfilled.

### Functions
|Purpose|Function|
|--|--|
|Creating a New Table (from arrays)|`NewTable = Table().with_columns(ColumnHeader1, ListOrArray1, ColumnHeader2, ListOrArray2, [etc, etc])` (it's much recommended to make a table using arrays as columns rather than using lists as rows.)|
|Creating a New Table (from .csv file)|`NewTable = Table.read_table('FileName')` (if the file in question is not in the same folder as the Exercise/Lab you're working on, this won't work. You can technically add the file path at the beginning of the quoted FileName, but it's MUCH easier to just make sure the Notebook page you're working on is in the same folder as the file by default.)|
|Revealing the full table|`ExistingTable.show(NumberOfRows)` use empty parentheses if you want all rows|
|Restricting a Table to certain parameters|`ExistingTable.where(ColumnHeader, are.Condition(Parameter))` (check below for conditions/parameters)|
|Selecting  row in Table format|`ExistingTable.take(IndexOfRow1, IndexOfRow2, [etc, etc])` (you can use functions to create arrays to put here for the row indices)|
|Selecting a column in Table format|`ExistingTable.select('ColumnHeader1' OR IndexOfColumn1, 'ColumnHeader2' OR IndexOfColumn2, [etc, etc])` (stick to one type; either all headers or all indices.) OR `existingTable[IndexOfColumn]`|
|Selecting a column in array format|`ExistingTable.column('ColumnHeader1' OR IndexOfColumn1, 'ColumnHeader2' OR IndexOfColumn2, [etc, etc])`|
|Adding one row to a Table|`ExistingTable = ExistingTable.with_row(NewRowAsList)`|
|Adding one column to a Table|`ExistingTable = ExistingTable.with_column(ColumnHeader, ListOrArray)`|
|Adding multiple rows to a Table|`ExistingTable = ExistingTable.with_rows(NewRowAsList1, NewRowAsList2, [etc, etc])`|
|Adding multiple columns to a Table|`ExistingTable = ExistingTable.with_columns(ColumnHeader1, ListOrArray1, ColumnHeader2, ListOrArray2, [etc, etc])`|
|Sorting the rows|`ExistingTable.sort(ColumnHeader, descending = True/False)` (descending is automatically set to false; put it to True if you want highest values first)|
|Getting the number of rows/columns|`ExistingTable.num_rows` OR `ExistingTable.num_columns`|
|Getting rid of certain columns|`ExistingTable.drop(UnwantedColumnHeader1, UnwantedColumnHeader2, [etc, etc])`|

### Conditions
|Purpose|Function (in np.where)|
|--|--|
|where target values are equal to the Target Element|`ExistingTable.where(ColumnHeader, are.equal_to(TargetElement))` OR `ExistingTable.where(ColumnHeader, TargetElement)`|
|where target values are NOT equal to the Target Element|`ExistingTable.where(ColumnHeader, are.not_equal_to(TargetElement))`|
|where target values are less than the Target Element|`ExistingTable.where(ColumnHeader, are.below(TargetElement))`|
|where target values are more than the Target Element|`ExistingTable.where(columnHeader, are.above(TargetElement))`|
|where target values are between two Target Elements|`ExistingTable.where(ColumnHeader, are.between(Element1, Element2))` (this includes values equal to Element1, but NOT those equal to Element2)|
|where target values contain the Target Element|`ExistingTable.where(ColumnHeader, are.containing(TargetElement))`|


One key detail is that you need to keep track of indexes. Taking the index of a table once gives you a column, and taking the index twice gives you a cell. For example, `existingTable[5][2]` goes to the table's 6th column (remember, indexes start counting at 0, not 1) and gets the cell of that column that is in the 3rd row.

Before we can even hope to work with tables, we must import the relevant libraries. These relevant libraries are the datascience module, which were will almost always import without its name.

In [1]:
from datascience import *
import numpy as np # this library is also used in these Exercises.

## Exercise 1: 
A central part of data manipulation is, well, the manipulation of data (well done Einstein). This primarily constitutes the use of FOM (Functions, Operations, and/or Methods) to show another aspect of data or to express existing data in a different format.

Why do we want to manipulate data?
* To remake it in a format that is more easily usable by the computer itself (Remember, your capabilities in analyzing data are determined at least partially by the capabilities of Jupyter itself!)
* To conduct statistical tests and other advanced actions (spoilers for Lab 05 and onwards)

The principal way to manipulate data is to extract it from the Table setting. While it may seem that Tables are the all-important focus of this class, don't fall for that illusion. **Tables are only a convenient way to organize and understand data. By themselves, they are incapable of actual analysis.** The true leaders here are the lists and arrays. I'll leave it to you to find the exact method that converts tables to arrays in the above table.

### 1.1: Title Length
Create the imdb Table from the file named 'imdb.csv', and extract the Title column as an array. 

In [2]:
imdb = Table.read_table('imdb.csv')
imdb

Votes,Rating,Title,Year,Decade
88355,8.4,M,1931,1930
132823,8.3,Singin' in the Rain,1952,1950
74178,8.3,All About Eve,1950,1950
635139,8.6,Léon,1994,1990
145514,8.2,The Elephant Man,1980,1980
425461,8.3,Full Metal Jacket,1987,1980
441174,8.1,Gone Girl,2014,2010
850601,8.3,Batman Begins,2005,2000
37664,8.2,Judgment at Nuremberg,1961,1960
46987,8,Relatos salvajes,2014,2010


In [3]:
titles = imdb.column('Title')

Use the relevant command, and the correct format, to get a list that contains the number of characters in each title. For example, the movie title 'León' would beget a value of 4 (it has 4 characters). Review the Exercises 02 if you do not know.

In [5]:
t_lengths = [len(x) for x in titles]
print(t_lengths)

[1, 19, 13, 4, 16, 17, 9, 13, 21, 16, 21, 11, 16, 20, 18, 13, 19, 28, 34, 21, 25, 19, 13, 36, 14, 12, 7, 12, 22, 9, 14, 4, 14, 21, 16, 11, 8, 18, 11, 14, 16, 19, 68, 19, 7, 20, 26, 24, 7, 10, 16, 13, 6, 24, 22, 5, 18, 38, 27, 16, 6, 11, 13, 13, 24, 8, 4, 13, 37, 19, 12, 12, 18, 12, 14, 32, 15, 9, 23, 9, 22, 9, 18, 12, 42, 20, 12, 10, 3, 13, 12, 22, 32, 23, 17, 6, 18, 26, 20, 35, 24, 16, 27, 10, 15, 12, 9, 8, 44, 18, 9, 14, 12, 16, 9, 15, 8, 54, 19, 15, 9, 11, 7, 13, 5, 11, 14, 14, 31, 10, 10, 8, 19, 18, 21, 7, 10, 7, 14, 6, 8, 45, 23, 16, 33, 21, 11, 16, 8, 14, 8, 7, 29, 26, 13, 42, 46, 18, 10, 5, 27, 17, 14, 13, 19, 17, 13, 7, 21, 5, 10, 5, 13, 7, 12, 22, 23, 17, 49, 12, 11, 11, 11, 2, 11, 22, 14, 15, 18, 9, 20, 17, 9, 9, 7, 7, 17, 12, 13, 13, 11, 16, 13, 34, 15, 13, 15, 17, 28, 13, 23, 21, 12, 21, 11, 13, 11, 11, 31, 6, 21, 5, 24, 31, 12, 7, 7, 12, 22, 9, 6, 8, 19, 18, 8, 19, 23, 23, 15, 19, 18, 20, 14, 12, 2, 12, 21, 8, 7, 37]


Append this column to the imdb table. Give it any column header you want. :']

In [6]:
imdb = imdb.with_column('Title Length', t_lengths)

What are the five movies with the longest names?

In [7]:
imdb.sort('Title Length', descending=True).show(5) # you don't have to do all this; any sort of showing the five movies is enough

Votes,Rating,Title,Year,Decade,Title Length
309141,8.5,Dr. Strangelove or: How I Learned to Stop Worrying and L ...,1964,1960,68
718645,8,Pirates of the Caribbean: The Curse of the Black Pearl,2003,2000,54
1099087,8.8,The Lord of the Rings: The Fellowship of the Ring,2001,2000,49
700283,8.7,Star Wars: Episode V - The Empire Strikes Back,1980,1980,46
1074146,8.9,The Lord of the Rings: The Return of the King,2003,2000,45


### 1.2: Multistep process
Add a column to the imdb table that shows how many a's that movie title has. For example, the movie title 'All About Eve' would have a value of 2.

I'll walk you through this one, since it does a good job of showing how multistep coding procedures work in regards to data manipulation.

In [8]:
imdb

Votes,Rating,Title,Year,Decade,Title Length
88355,8.4,M,1931,1930,1
132823,8.3,Singin' in the Rain,1952,1950,19
74178,8.3,All About Eve,1950,1950,13
635139,8.6,Léon,1994,1990,4
145514,8.2,The Elephant Man,1980,1980,16
425461,8.3,Full Metal Jacket,1987,1980,17
441174,8.1,Gone Girl,2014,2010,9
850601,8.3,Batman Begins,2005,2000,13
37664,8.2,Judgment at Nuremberg,1961,1960,21
46987,8,Relatos salvajes,2014,2010,16


Step 1: You may have noticed that the letter A sometimes appears as uppercase (A) and other times as lowercase (a). Python treats these two as totally distinct characters (like apples and oranges). Extract the 'Title' column as an array again (or use the same variable if you already defined it in Exercise 1.1) and use the method `String.lower()` to convert all characters to lowercase. Hmm, how do you apply a method to all elements in a list?

In [9]:
lowercase_titles = [x.lower() for x in titles]
print(lowercase_titles)

['m', "singin' in the rain", 'all about eve', 'léon', 'the elephant man', 'full metal jacket', 'gone girl', 'batman begins', 'judgment at nuremberg', 'relatos salvajes', 'it happened one night', 'gran torino', 'some like it hot', 'inglourious basterds', 'the princess bride', 'mononoke-hime', 'saving private ryan', 'mr. smith goes to washington', 'butch cassidy and the sundance kid', 'das leben der anderen', 'in the name of the father', 'the grapes of wrath', 'the godfather', 'bom yeoreum gaeul gyeoul geurigo bom', 'v for vendetta', 'mary and max', 'warrior', 'intouchables', 'la battaglia di algeri', 'the sting', 'shutter island', 'jaws', 'before sunrise', '2001: a space odyssey', '12 years a slave', 'the killing', 'papillon', 'lawrence of arabia', 'the shining', 'cool hand luke', 'a beautiful mind', 'inside out (2015/i)', 'dr. strangelove or: how i learned to stop worrying and love the bomb', 'paris, texas (1984)', 'snatch.', 'the bourne ultimatum', 'x-men: days of future past', 'the s

Step 2: Now comes the heart of the problem, that is, how to make Python count all the a's in a title. There are many ways to go about this, and you are welcome to create your own solution here. However, one of the simplest I know revolves around the `String.split(splitBoundary)` method, which was introduced in the Exercises for Lab 02. Refer back there (particularly to the 2.3 and 2.4 exercise problems) or to your own toolbox (that you *should* create) for how to use that method.

Note that, in the example below, for each instance of A, the method effectively cleaves the string at that location. Thus, you can expect that a word with one A begets two lists when put in the .split method, two A's begets three lists, and so forth.

Your lightbulb may already be alight about where I'm going with this, but I'll take it slow in case you're still confused. Use .split to get the various cleavings of each title. This will be a list of lists.

In [10]:
# examples!
print('chameleon'.split('a')) # one a begets two fragments
print('bargain'.split('a')) # two a begets three fragments
print('avalanche'.split('a')) # three a begets four fragments (note that the .split includes a separate fragment even though a is the first letter!)

# in a list:
example_words = make_array('chameleon', 'bargain', 'avalanche') # this represents the array of titles
split_list = [x.split('a') for x in example_words] # this is a list of lists
split_list

['ch', 'meleon']
['b', 'rg', 'in']
['', 'v', 'l', 'nche']


[['ch', 'meleon'], ['b', 'rg', 'in'], ['', 'v', 'l', 'nche']]

In [12]:
# Use the examples to get the various fragments of the titles in your title column array
split_a = [x.split('a') for x in lowercase_titles]
print(split_a)

[['m'], ["singin' in the r", 'in'], ['', 'll ', 'bout eve'], ['léon'], ['the eleph', 'nt m', 'n'], ['full met', 'l j', 'cket'], ['gone girl'], ['b', 'tm', 'n begins'], ['judgment ', 't nuremberg'], ['rel', 'tos s', 'lv', 'jes'], ['it h', 'ppened one night'], ['gr', 'n torino'], ['some like it hot'], ['inglourious b', 'sterds'], ['the princess bride'], ['mononoke-hime'], ['s', 'ving priv', 'te ry', 'n'], ['mr. smith goes to w', 'shington'], ['butch c', 'ssidy ', 'nd the sund', 'nce kid'], ['d', 's leben der ', 'nderen'], ['in the n', 'me of the f', 'ther'], ['the gr', 'pes of wr', 'th'], ['the godf', 'ther'], ['bom yeoreum g', 'eul gyeoul geurigo bom'], ['v for vendett', ''], ['m', 'ry ', 'nd m', 'x'], ['w', 'rrior'], ['intouch', 'bles'], ['l', ' b', 'tt', 'gli', ' di ', 'lgeri'], ['the sting'], ['shutter isl', 'nd'], ['j', 'ws'], ['before sunrise'], ['2001: ', ' sp', 'ce odyssey'], ['12 ye', 'rs ', ' sl', 've'], ['the killing'], ['p', 'pillon'], ['l', 'wrence of ', 'r', 'bi', ''], ['th

Step 3: Use another function in Exercises 02 to get the length of each sub-list in the list. Adjust the values so that they match the number of a's. (Remember, a list of fragments of length 3 came from a title that had TWO a's, not three!)

In [13]:
# example! A function on all sublists of a list. This function is NOT what you want for step 3, however.
[x[0] for x in split_list] # this takes the first term of all the sublists.

['ch', 'b', '']

In [15]:
num_a = [len(x)-1 for x in split_a]
print(num_a)

[0, 1, 2, 0, 2, 2, 0, 2, 1, 3, 1, 1, 0, 1, 0, 0, 3, 1, 3, 2, 2, 2, 1, 1, 1, 3, 1, 1, 5, 0, 1, 1, 0, 2, 3, 0, 1, 4, 0, 1, 2, 0, 3, 2, 1, 1, 2, 1, 0, 0, 0, 2, 1, 2, 0, 1, 2, 1, 3, 2, 1, 0, 0, 1, 2, 1, 1, 2, 0, 2, 1, 1, 2, 1, 2, 3, 1, 3, 3, 0, 0, 0, 0, 2, 2, 3, 0, 0, 1, 1, 1, 2, 3, 4, 2, 1, 1, 2, 2, 3, 2, 1, 1, 1, 3, 1, 0, 1, 5, 1, 2, 1, 2, 2, 0, 0, 2, 5, 1, 1, 0, 1, 1, 1, 2, 0, 0, 1, 3, 1, 0, 0, 3, 2, 0, 0, 2, 1, 0, 1, 2, 0, 1, 1, 5, 1, 1, 0, 1, 0, 0, 1, 2, 2, 2, 2, 3, 2, 4, 1, 0, 0, 3, 1, 0, 1, 0, 1, 0, 0, 2, 0, 0, 0, 1, 2, 1, 1, 0, 1, 0, 1, 2, 0, 1, 1, 1, 0, 0, 2, 1, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 0, 0, 5, 3, 0, 1, 0, 1, 1, 3, 1, 0, 1, 0, 1, 0, 1, 2, 0, 3, 0, 4, 0, 0, 2, 1, 0, 2, 1, 1, 1, 3, 3, 0, 0, 2, 0, 0, 2, 2, 3, 1, 1, 0, 0, 3, 0, 0, 1]


Step 4: Append the column to the imdb table.

In [16]:
imdb = imdb.with_column('A count', num_a)

What are the five movies with the most A's in the title?

In [17]:
imdb.sort('A count', descending=True).show(5)

Votes,Rating,Title,Year,Decade,Title Length,A count
32385,8,La battaglia di Algeri,1966,1960,22,5
466111,8,Harry Potter and the Deathly Hallows: Part 2,2011,2010,44,5
718645,8,Pirates of the Caribbean: The Curse of the Black Pearl,2003,2000,54,5
59377,8,Lagaan: Once Upon a Time in India,2001,2000,33,5
453748,8.3,Indiana Jones and the Last Crusade,1989,1980,34,5


### 1.3: Date Formatting
Now time for some actually-real-world applications. A very common sight in Tables and Data Statistics in general is the reformatting of dates to render them more code-oriented. After all, Python has no idea what the slashes in '7/31/2025' mean, nor does it comprehend the colons in '6:30:25'. To be able to work with times and dates, we need to reexpress these values in terms of a single unit (e.g. hours) instead of multiple units (e.g. hours : minutes : seconds).

Consider the times in the following table. Our task in this exercise is to convert them from hh:mm:ss to some amount of hours past midnight. For example, a time of 11:30:00 gets a value of 11.5 hours past midnight, and 7:25:52 gets a value of 7.431 hours past midnight. In essence, we are doing the following: (using 15:35:20 as an example)
1. Getting the seconds in terms of a decimal portion of a minute (e.g. 20 seconds is 1/3 of a minute or 0.33 minutes)
2. Adding that to the minute count (35+0.33 = 35.33 minutes)
3. Expressing that minute count in terms of a decimal portion of an hour (35.33 minutes becomes 0.589 hours)
4. Adding that to the hours (15+0.589 = 15.589 hours)

In [18]:
unformatted = ["4:50:31", "7:12:45", "9:38:09", "11:27:56", "13:03:22", "15:41:10", "17:19:44", "20:02:37", "22:46:05", "00:15:28"]

times = Table().with_columns('hh:mm:ss times', unformatted)
times # this is typically what these times will look like within a table. 
# Of course, other data will also be there, but for the sake of simplicity, this is not included.

hh:mm:ss times
4:50:31
7:12:45
9:38:09
11:27:56
13:03:22
15:41:10
17:19:44
20:02:37
22:46:05
00:15:28


When having to format a whole list, it's a good idea to focus solely on one singular value, and then extrapolate to the entire thing. Try working with just this first value below.

In [19]:
test_time = times[0][0]
test_time

'4:50:31'

Here are some guiding questions:
* How do you 'unpack' the values in the data format?
* Are these unpacked values able to be worked with numerically? What do you have to do to make them so?
* Once you have have worked it out and gotten the singular test_time to work, what do you have to do to apply this to the whole table column?
* Remember, what first step do you have to do to make data within tables open to analysis?

Once you've solved it, append the formatted times as another column to the table.

In [20]:
# working with the singular case
unpacking = test_time.split(':')
hour = float(unpacking[0])
minute = float(unpacking[1])
second = float(unpacking[2])
hour_time = hour + minute/60 + second/3600
hour_time

4.841944444444444

In [22]:
# translating to the list
original = times.column('hh:mm:ss times')
unpack_list = [x.split(':') for x in original]
hour_list = [float(x[0]) + float(x[1])/60 + float(x[2])/3600 for x in unpack_list]
hour_list

[4.841944444444444,
 7.2125,
 9.635833333333332,
 11.465555555555556,
 13.056111111111111,
 15.686111111111112,
 17.328888888888887,
 20.04361111111111,
 22.768055555555556,
 0.2577777777777778]

In [23]:
times = times.with_column('Hours', hour_list)
times

hh:mm:ss times,Hours
4:50:31,4.84194
7:12:45,7.2125
9:38:09,9.63583
11:27:56,11.4656
13:03:22,13.0561
15:41:10,15.6861
17:19:44,17.3289
20:02:37,20.0436
22:46:05,22.7681
00:15:28,0.257778


## Exercise 2:
A crucial part of data science is being able to translate abstract questions into code lines that will give you the answer to the question. This type of translation is akin to mathematical calculations or finding a route to drive to a certain place. You've already seen such questions in the lab itself, but given how important they are to your projects, you will need plenty more practice for this. The main strategy is to know what your answer should look like and how to get there, and therefore to design a series of steps to get from the entire table to your final destination.

### 2.1: World Populations
These questions regard the dataset 'world_populations.csv'. I've done some prep work below to include the years.

In [24]:
pops = Table.read_table('world_population.csv')
pops = pops.with_column('Year', np.arange(1950, 2016))
pops

Population,Year
2557628654,1950
2594939877,1951
2636772306,1952
2682053389,1953
2730228104,1954
2782098943,1955
2835299673,1956
2891349717,1957
2948137248,1958
3000716593,1959


How many years did the world spend with a population between 4 and 5 billion?

In [26]:
pops.where('Population', are.between(4000000000, 5000000000)).num_rows

13

Has the world population grown a larger amount from 1950 to 2000 or from 2000 to 2016? By how much are these amounts different?

In [29]:
old = pops.where('Year', are.between(1950,2000))
new = pops.where('Year', are.between(2000,2016))
old_growth = old.column('Population')[-1]-old.column('Population')[0]
new_growth = new.column('Population')[-1]-new.column('Population')[0]
print(old_growth, new_growth)
print('The period from 1950 to 2000 saw a greater increase, as the population grew', old_growth-new_growth, 'more than it did between 2000 and 2016.')

3454446268 1167918628
The period from 1950 to 2000 saw a greater increase, as the population grew 2286527640 more than it did between 2000 and 2016.


### 2.2: Farmers Markets
These questions regard the dataset 'farmers_markets.csv'.

In [30]:
farmers_markets = Table().read_table('farmers_markets.csv')
farmers_markets = farmers_markets.drop('FMID','Website','Facebook','Twitter','Youtube','OtherMedia','Season1Date','Season1Time','Season2Date','Season2Time','Season3Date','Season3Time','Season4Date','Season4Time')
farmers_markets

MarketName,street,city,County,State,zip,x,y,Location,Credit,WIC,WICcash,SFMNP,SNAP,Organic,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,Tofu,WildHarvested,updateTime
Caledonia Farmers Market Association - Danville,nan,Danville,Caledonia,Vermont,05828,-72.1403,44.411,nan,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,Y,Y,Y,Y,N,Y,Y,Y,N,Y,N,Y,N,N,6/28/2016 12:10:09 PM
Stearns Homestead Farmers' Market,6975 Ridge Road,Parma,Cuyahoga,Ohio,44130,-81.7286,41.3751,nan,Y,Y,N,Y,Y,-,Y,N,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,Y,N,N,N,N,N,N,N,Y,N,N,N,Y,N,N,4/9/2016 8:05:17 PM
100 Mile Market,507 Harrison St,Kalamazoo,Kalamazoo,Michigan,49007,-85.5749,42.296,nan,Y,Y,N,Y,Y,N,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,Y,N,N,Y,Y,N,N,N,N,N,4/16/2016 12:37:56 PM
106 S. Main Street Farmers Market,106 S. Main Street,Six Mile,nan,South Carolina,29682,-82.8187,34.8042,nan,Y,N,N,N,N,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2013
10th Steet Community Farmers Market,10th Street and Poplar,Lamar,Barton,Missouri,64759,-94.2746,37.4956,nan,Y,N,N,N,N,-,Y,N,Y,N,Y,N,Y,Y,Y,Y,N,Y,N,N,Y,Y,Y,Y,N,N,N,N,Y,N,N,N,N,N,N,10/28/2014 9:49:46 AM
112st Madison Avenue,112th Madison Avenue,New York,New York,New York,10029,-73.9493,40.7939,Private business parking lot,N,N,Y,Y,N,-,Y,N,Y,Y,N,N,Y,Y,Y,Y,N,N,N,Y,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,3/1/2012 10:38:22 AM
12 South Farmers Market,3000 Granny White Pike,Nashville,Davidson,Tennessee,37204,-86.7907,36.1184,nan,Y,N,N,N,Y,Y,Y,Y,N,Y,Y,N,Y,Y,Y,Y,Y,Y,N,N,N,Y,Y,Y,N,N,Y,N,Y,N,Y,Y,Y,N,N,5/1/2015 10:40:56 AM
125th Street Fresh Connect Farmers' Market,"163 West 125th Street and Adam Clayton Powell, Jr. Blvd.",New York,New York,New York,10027,-73.9482,40.809,Federal/State government building grounds,Y,Y,N,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,Y,Y,Y,N,Y,Y,N,Y,N,Y,N,N,N,N,4/7/2014 4:32:01 PM
12th & Brandywine Urban Farm Market,12th & Brandywine Streets,Wilmington,New Castle,Delaware,19801,-75.5345,39.7421,"On a farm from: a barn, a greenhouse, a tent, a stand, etc",N,N,N,N,Y,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,4/3/2014 3:43:31 PM
14&U Farmers' Market,1400 U Street NW,Washington,District of Columbia,District of Columbia,20009,-77.0321,38.917,Other,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,N,Y,Y,Y,Y,N,Y,N,Y,Y,Y,N,N,N,N,N,Y,Y,Y,Y,N,N,N,N,4/5/2014 1:49:04 PM


What percentage of these markets have 'Farmers' in their name?

In [32]:
farmers_named = farmers_markets.where('MarketName', are.containing('Farmers'))
farmers_named.num_rows/farmers_markets.num_rows

0.7502925345190733

What markets lie within 1 degree, latitude and longitude, of the Temple University's SERC building (the coordinates are -75.153, 39.982)? For reference, one degree of latitude/longitude is equal to 69 miles. (nice)

In [34]:
farmers_markets.where('x', are.between(-76.153,-74.153)).where('y', are.between(38.982,40.982)) # they don't have to list all 250-so names.

MarketName,street,city,County,State,zip,x,y,Location,Credit,WIC,WICcash,SFMNP,SNAP,Organic,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,Tofu,WildHarvested,updateTime
12th & Brandywine Urban Farm Market,12th & Brandywine Streets,Wilmington,New Castle,Delaware,19801,-75.5345,39.7421,"On a farm from: a barn, a greenhouse, a tent, a stand, etc",N,N,N,N,Y,N,N,N,N,N,N,N,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,4/3/2014 3:43:31 PM
18th and Christian Farmers' Market,18th Street and Christian Street,Philadelphia,Philadelphia,Pennsylvania,19146,-75.1734,39.9409,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 3:33:48 PM
22nd and Tasker Farmers' Market,22nd and Tasker Streets,Philadelphia,Philadelphia,Pennsylvania,19146,-75.1813,39.9323,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:29:42 PM
26th and Allegheny,26th Street and W Allegheny Avenue,Philadelphia,Philadelphia,Pennsylvania,19129,-75.1723,40.0041,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 3:17:20 PM
29th and Wharton Farmers' Market,29th and Wharton Streets,Philadelphia,Philadelphia,Pennsylvania,19146,-75.192,39.9373,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:47:10 PM
33rd and Diamond Farmers' Market,N 33rd and Diamond Streets,Philadelphia,Philadelphia,Pennsylvania,19121,-75.1872,39.9888,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:28:01 PM
4th and Lehigh Farmers' Market,N 4th Street and W. Lehigh Avenue,Philadelphia,Philadelphia,Pennsylvania,19133,-75.1387,39.9917,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 3:11:40 PM
52nd and Haverford Farmers' Market,N 52nd Street and Haverford Avenue,Philadelphia,Philadelphia,Pennsylvania,19139,-75.2245,39.9651,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:38:14 PM
58th and Chester Farmers' Market,58th Street and Chester Avenue,Philadelphia,Philadelphia,Pennsylvania,19143,-75.2283,39.9357,nan,Y,Y,N,Y,Y,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,Y,N,N,N,N,N,N,3/24/2015 2:35:04 PM
ABV Farm Market,690 Allegheny Road,Mount Bethe,nan,Pennsylvania,18343,-75.0998,40.8743,nan,Y,N,N,N,N,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,2013


What markets sell Meat and Tofu but no vegetables?

In [35]:
farmers_markets.where('Meat', 'Y').where('Tofu', 'Y').where('Vegetables', 'N')

MarketName,street,city,County,State,zip,x,y,Location,Credit,WIC,WICcash,SFMNP,SNAP,Organic,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,Tofu,WildHarvested,updateTime
Old Town Alexandria Farmers' Market,301 King Street,Alexandria,Alexandria,Virginia,22314,-77.043,38.8048,nan,Y,N,N,N,Y,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,Y,N,Y,Y,N,Y,N,7/11/2015 6:34:15 AM
Saratoga Certified Farmers' Market,"West Valley College, 14000 Fruitvale Avenue",Saratoga,Santa Clara,California,95070,-122.011,37.2642,Federal/State government building grounds,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,N,Y,Y,N,Y,N,Y,Y,Y,Y,Y,Y,N,Y,Y,Y,Y,Y,Y,N,Y,N,4/1/2014 4:43:25 PM
Valley Farmers' Market Association LLC,"3201 ""F"" Street",Bakersfield,Kern,California,nan,-119.025,35.3886,Private business parking lot,Y,Y,N,Y,Y,Y,Y,Y,Y,Y,Y,N,Y,N,Y,Y,N,Y,N,Y,Y,N,Y,Y,Y,N,Y,N,Y,N,Y,N,Y,Y,N,4/1/2014 9:41:05 AM


## Exercise 3:
### 3.1:
In this section, we will introduce to ourselves a very useful function whose name I will not spoil yet. This involves the use of the farmers_markets table. If you look at the bunch of columns at the far right end, most of them show whether or not a given market sells a particular good. A 'Y' means that the market does sell it, and 'N' means that it doesn't.

Suppose you pick the Cheese column (because I like cheese) that has these Ys and Ns and want to know how many of them have Ys (how many sell cheese) and how many of them have Ns (how many don't sell cheese). (Note that Ys and Ns are the only inputs in the column.) How do you do this?

Step 1: Define the `Y_cheese` variable as the farmers_markets table containing only the rows with 'Y' in the Cheese column, and the `N_cheese` variable as the table containing only the rows with 'N' in the Cheese column.

In [36]:
Y_cheese = farmers_markets.where('Cheese', 'Y')
N_cheese = farmers_markets.where('Cheese', 'N')

Step 2: Get the number of rows in the Y_cheese and N_cheese, and define them as `Y_rows` and `N_rows`.

In [38]:
Y_rows = Y_cheese.num_rows
N_rows = N_cheese.num_rows

Note that what you've gotten right now, in Y_rows and N_rows, is the number of markets that have cheese (number of rows that have 'Y') and don't have cheese (number of rows that have 'N'). Now, assuming you've done your duty in properly defining these variables, I'll package this information in a format that is much more pleasant to your eyes.

In [39]:
# don't change this cell!
inputs = ['Y', 'N']
number_rows = [Y_rows, N_rows]
summary_cheese = Table().with_columns('Cheese', inputs, 'count', number_rows)
summary_cheese = summary_cheese.sort('Cheese')
summary_cheese

Cheese,count
N,5802
Y,2744


This is the essence of our mystery function: for each unique element in a given column, it will tell you how many rows in the table have that input for that column. Here is that mystery function below.

In [40]:
farmers_markets.group('Cheese')

Cheese,count
N,5802
Y,2744


### 3.2: Practice with `.group`
Which state has the most markets?

In [44]:
farmers_markets.group('State').sort('count', descending=True).show(1)

State,count
California,755


Going back to the imdb table, which decade had the most movies?

In [45]:
imdb.group('Decade').sort('count', descending=True).show(1)

Decade,count
2000,50


What's the maximum number of A's a movie had? Create the .group table and assign it the variable `A_stats`.

Note that, since the .group method gives you a table, you can modify that table in itself. Use this fact to add on to `A_stats` another column that tells you what percentage of the total movies had that number of A's. (e.g. What percent had zero A's? What percent had one A? and so forth)

In [48]:
A_stats = imdb.group('A count')
A_stats

A count,count
0,81
1,84
2,51
3,25
4,4
5,5


In [49]:
A_count = A_stats.column('count')
A_stats = A_stats.with_column('percentage', [x/imdb.num_rows for x in A_count])
A_stats

A count,count,percentage
0,81,0.324
1,84,0.336
2,51,0.204
3,25,0.1
4,4,0.016
5,5,0.02


### 3.3: `.group` with a function
When paired with a function as the second argument, the resultant table is very different. Let's look back at the imdb table, and apply the .group method and the np.average function to the 'Decade' column. It does filter the table according to each unique input of Decade, but instead of taking the number of rows for each input in Decade, it looks at the other columns besides Decade and takes their average (applies the specified function).

In [50]:
imdb.group('Decade', np.average)

Decade,Votes average,Rating average,Title average,Year average,Title Length average,A count average
1920,64854,8.25,,1924.75,10.25,0.25
1930,125825,8.27143,,1935.57,15.2857,0.428571
1940,122767,8.25714,,1944.07,16.7143,1.35714
1950,109829,8.23333,,1955,15.0667,0.833333
1960,147391,8.23182,,1964.14,21.1364,1.81818
1970,353377,8.30952,,1975.62,12.8571,1.09524
1980,295290,8.24194,,1984.29,16.1935,1.74194
1990,559449,8.35714,,1995.19,14.5238,0.952381
2000,481929,8.26,,2004.26,18.44,1.14
2010,438006,8.2,,2012.31,16.1034,1.37931


Let's take the first row as an example. 

What it tells us is that, for all the 1920 movies in the table...
* They got 64854 votes on average
* They had an average rating of 8.25
* The average year in which they were released was 1924 (ignoring decimals)
* They had an average of 0.25 A's in their name.

And this goes for all the movies released in the decade of 1920. The 1930 movies' average stats are in the second row, the 40s' stats in the third, and so forth. 

Note that the Title column is blank because you can't take the average of non-numerical values.

**What functions can you use in `.group`?**

Any function works SO LONG AS its only input is a list or array. Methods do not work, as the .group assumes that this only input is in the parentheses following the function name, and the method's input is before it.

Assume that you have already done 1.2 and thus have the 'Number of As' column at your disposal. If you're stuck on some of these, you are more than welcome to create extra code lines to study the relevant data table or make your steps a bit more coherent.

What's the total number of A's in movies of the 1940s?

In [56]:
imdb.group('Decade', sum).where('Decade', 1940)

Decade,Votes sum,Rating sum,Title sum,Year sum,Title Length sum,A count sum
1940,1718735,115.6,,27217,234,19


What decade had the longest-titled movie names?

In [58]:
imdb.group('Decade', max).sort('Title Length max', descending=True).show(1)

Decade,Votes max,Rating max,Title max,Year max,Title Length max,A count max
1960,447875,8.9,Yôjinbô,1969,68,5


What's the westernmost market that sells Eggs?

In [60]:
farmers_markets.group('Eggs', min).sort('Eggs', descending=True).show(1)

Eggs,MarketName min,street min,city min,County min,State min,zip min,x min,y min,Location min,Credit min,WIC min,WICcash min,SFMNP min,SNAP min,Organic min,Bakedgoods min,Cheese min,Crafts min,Flowers min,Seafood min,Herbs min,Vegetables min,Honey min,Jams min,Maple min,Meat min,Nursery min,Nuts min,Plants min,Poultry min,Prepared min,Soap min,Trees min,Wine min,Coffee min,Beans min,Fruits min,Grains min,Juices min,Mushrooms min,PetFood min,Tofu min,WildHarvested min,updateTime min
Y,Caledonia Farmers Market Association - Danville,1102 McConnell Road,New Canaan,Abbeville,Alabama,0033,-159.396,18.0093,Closed-off public street,N,N,N,N,N,-,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,1/1/2015 10:12:55 PM


In [64]:
farmers_markets.where('x', are.between(-159.397,-159.395)) # the omission of digits makes this very finnicky

MarketName,street,city,County,State,zip,x,y,Location,Credit,WIC,WICcash,SFMNP,SNAP,Organic,Bakedgoods,Cheese,Crafts,Flowers,Eggs,Seafood,Herbs,Vegetables,Honey,Jams,Maple,Meat,Nursery,Nuts,Plants,Poultry,Prepared,Soap,Trees,Wine,Coffee,Beans,Fruits,Grains,Juices,Mushrooms,PetFood,Tofu,WildHarvested,updateTime
Kauai Community Market,3-1901 Kaumualii Highway,Lihue,Kauai,Hawaii,96766,-159.396,21.967,Educational institution,Y,N,N,N,Y,Y,Y,Y,N,Y,Y,N,Y,Y,Y,Y,N,N,N,N,Y,N,Y,Y,N,N,N,N,Y,N,N,N,N,N,N,6/12/2013 1:26:01 PM


## Exercise 4: Pivot
A function that is similar to .group is .pivot, where the resultant table denotes how many rows have TWO indicated elements in TWO columns. Take a look at the table formed by this following cell.

In [65]:
imdb.pivot('Rating', 'Decade')

Decade,8.0,8.1,8.2,8.3,8.4,8.5,8.6,8.7,8.8,8.9,9.0,9.2
1920,0,0,2,2,0,0,0,0,0,0,0,0
1930,0,3,1,0,1,2,0,0,0,0,0,0
1940,2,3,1,4,2,1,1,0,0,0,0,0
1950,6,7,5,5,3,2,0,1,0,1,0,0
1960,6,4,2,6,0,2,1,0,0,1,0,0
1970,5,5,1,4,0,2,0,2,0,0,1,1
1980,5,6,6,4,7,2,0,1,0,0,0,0
1990,9,4,5,5,3,4,5,3,1,2,0,1
2000,8,14,8,5,4,6,1,1,1,2,0,0
2010,9,7,3,3,2,3,1,1,0,0,0,0


Looking at the first row, it tells us that, among the 1920 movies, 2 of them scored an 8.2. That is, there were two rows that simultaneously had a value of 1920 for the decade and 8.2 for the rating. 

Note the input order; the column elements on the headers correspond to the FIRST input, and those on the side axis correspond to the SECOND input. Remember, columns are much more workable than rows when it comes to data manipulation.

### 4.1: 
Which state has the largest *proportion* of markets that sell seafood? Before you start, think about this question and come up with a hypothesis.

Hypothesis: Those on the coast, like Rhode Island, California, Florida, will likely have the highest proportions.

Make a pivot table and add necessary adjustments to display the proportions. How do you get the proportions using your given data? You may need the Table.rows features, which groups into a list-like format all the rows.

In [82]:
seafood_state = farmers_markets.pivot('Seafood', 'State') # this MAY be challenging, I'll be honest. 
seafood_state.rows

Rows(State                | N    | Y
Alabama              | 138  | 4
Alaska               | 29   | 9
Arizona              | 74   | 25
Arkansas             | 98   | 4
California           | 560  | 195
Colorado             | 121  | 35
Connecticut          | 140  | 16
Delaware             | 29   | 6
District of Columbia | 28   | 8
Florida              | 160  | 86
... (43 rows omitted))

In [83]:
seafood_percent = [x[2]/(x[1]+x[2]) for x in seafood_state.rows]
seafood_state = seafood_state.with_column('Percentage', seafood_percent)
seafood_state

State,N,Y,Percentage
Alabama,138,4,0.028169
Alaska,29,9,0.236842
Arizona,74,25,0.252525
Arkansas,98,4,0.0392157
California,560,195,0.258278
Colorado,121,35,0.224359
Connecticut,140,16,0.102564
Delaware,29,6,0.171429
District of Columbia,28,8,0.222222
Florida,160,86,0.349593


In [84]:
seafood_state.sort('Percentage', descending=True)

State,N,Y,Percentage
Massachusetts,205,111,0.351266
Florida,160,86,0.349593
Washington,115,59,0.33908
Rhode Island,40,19,0.322034
Louisiana,53,25,0.320513
New Hampshire,67,28,0.294737
Oregon,120,50,0.294118
California,560,195,0.258278
Arizona,74,25,0.252525
Alaska,29,9,0.236842


### 4.2:
What Market Location will most often accept credit cards?

In [87]:
credit_location = farmers_markets.pivot('Credit', 'Location')
credit_location.rows

Rows(Location                                                     | N    | Y
Closed-off public street                                     | 101  | 205
Co-located with wholesale market facility                    | 1    | 5
Educational institution                                      | 34   | 77
Faith-based institution (e.g., church, mosque, synagogue ... | 56   | 71
Federal/State government building grounds                    | 19   | 58
Healthcare Institution                                       | 40   | 41
Local government building grounds                            | 483  | 556
On a farm from: a barn, a greenhouse, a tent, a stand, etc   | 27   | 64
Other                                                        | 215  | 435
Private business parking lot                                 | 351  | 437
... (1 rows omitted))

In [88]:
credit_percent = [x[2]/(x[1]+x[2]) for x in credit_location.rows]
credit_location = credit_location.with_column('Percentage', credit_percent)
credit_location

Location,N,Y,Percentage
Closed-off public street,101,205,0.669935
Co-located with wholesale market facility,1,5,0.833333
Educational institution,34,77,0.693694
"Faith-based institution (e.g., church, mosque, synagogue ...",56,71,0.559055
Federal/State government building grounds,19,58,0.753247
Healthcare Institution,40,41,0.506173
Local government building grounds,483,556,0.53513
"On a farm from: a barn, a greenhouse, a tent, a stand, etc",27,64,0.703297
Other,215,435,0.669231
Private business parking lot,351,437,0.554569


In [89]:
credit_location.sort('Percentage', descending=True)

Location,N,Y,Percentage
Co-located with wholesale market facility,1,5,0.833333
Federal/State government building grounds,19,58,0.753247
"On a farm from: a barn, a greenhouse, a tent, a stand, etc",27,64,0.703297
Educational institution,34,77,0.693694
Closed-off public street,101,205,0.669935
Other,215,435,0.669231
"Faith-based institution (e.g., church, mosque, synagogue ...",56,71,0.559055
Private business parking lot,351,437,0.554569
Local government building grounds,483,556,0.53513
Healthcare Institution,40,41,0.506173


## Conclusion
And that's the end of Exercises 03! I admit, this was a beefy one. Please provide feedback and any avenues for improvement in the cell below.

Unfortunately, I didn't put any secret codes in this Exercise 03. (Gotta make all five points really worth your while!)